src/Haskell/Ide/Engine/Plugin/Ormolu.hs

```diff
@@ -34,24 +40,73 @@ ormoluDescriptor plId = PluginDescriptor
 
 
 provider :: FormattingProvider
-provider _contents _uri _typ _opts =
 #if __GLASGOW_HASKELL__ >= 806
-  case _typ of
-    FormatRange _ -> return $ IdeResultFail (IdeError PluginError (T.pack "Selection formatting for Ormolu is not currently supported.") Null)
-    FormatText -> pluginGetFile _contents _uri $ \file -> do
-        opts <- lookupComponentOptions file
-        let opts' = map DynOption $ filter exop $ join $ maybeToList $ componentOptions <$> opts
-            conf  = Config opts' False False True False
-        result <- liftIO $ try @OrmoluException (ormolu conf file (T.unpack _contents))
-
-        case result of
-          Left  err -> return $ IdeResultFail (IdeError PluginError (T.pack $ "ormoluCmd: " ++ show err) Null)
-          Right new -> return $ IdeResultOk [TextEdit (fullRange _contents) new]
-  where
-    exop s =
-      "-X" `isPrefixOf` s
-      || "-fplugin=" `isPrefixOf` s
-      || "-pgmF=" `isPrefixOf` s
+{-# LANGUAGE BlockArguments #-}
+provider contents uri typ _ = pluginGetFile contents uri $ \fp -> do
+  opts <- lookupComponentOptions fp
+  let opts' =
+        map DynOption
+          $   filter exop
+          $   join
+          $   maybeToList
+          $   componentOptions
+          <$> opts
+      conf = Config opts' False False True False
+      fmt :: T.Text -> IdeM (Either OrmoluException T.Text)
+      fmt cont = liftIO $ try @OrmoluException (ormolu conf fp $ T.unpack cont)
+
+  case typ of
+    FormatText -> ret (fullRange contents) <$> fmt contents
+    FormatRange r ->
+      let
+        txt = T.lines $ extractRange r contents
+        lineRange (Range (Position sl _) (Position el _)) =
+          Range (Position sl 0) $ Position el $ T.length $ last txt
+        -- Pragmas will not be picked up in a non standard location,
+        -- or when range starts on a Pragma
+        extPragmas = takeWhile ("{-#" `T.isPrefixOf`)
```

Comment:
Given we are working in the context of a module that is likely to already be loaded by GHC, in principle we should be able to query these directly.

Reply:
And if we put this behaviour into a utility, we can either get the pragmas from the module `DynFlags`, or do this parsing as a fallback. And the GHC API already has a function to do that.

Reply:
I didn't see that I could show extensions from the `EnumSet` in `DynFlags`, I mistakenly thought I was going to have to test if each extension was set, to construct the list.